### ISSUE TITLE
The constraint ... is no smaller than the instance head... (Use UndecidableInstances ...)

### ISSUE THREAD

**Issue by silky** [haskell/error-messages#55]
Example:

```
{-# language FlexibleInstances #-}

module Main where

class X x where
  x :: x

class Y y where
  y :: y

instance X a => Y a where
  y = undefined
```

The error is:

```
/home/noon/tmp/err/src/Main.hs:11:10: error:
    • The constraint ‘X a’ is no smaller than the instance head ‘Y a’
      (Use UndecidableInstances to permit this)
    • In the instance declaration for ‘Y a’
   |
11 | instance X a => Y a where
```

Of course, `UndecidableInstances` is perhaps one fix, but a "better" one, as I understand it, [is the newtype wrapper](https://stackoverflow.com/a/7198951).

**Comment by goldfirere** [comment]
I personally prefer `UndecidableInstances` as the fix. I've never quite understood why Haskell is tetchy around compiler termination. When we write a non-terminating run-time program, we expect running that program not to terminate. When we write a non-terminating compile-time program, we should expect compiling that program not to terminate, too. Of course, it's better when we get an error message. But nothing really goes wrong if the compiler fails to terminate. In practice, GHC strives (and succeeds, in my experience) to generate an error message saying that it's not making any progress, instead of actually just hanging.

In this particular case, adding a newtype wrapper has potentially large consequences, requiring the programmer to add newtype injections and projections throughout their code. On the other hand, `-XUndecidableInstances` is quite easy to apply, and the only real downside is that non-terminating compile-time programs will get a slightly harder-to-understand error message (it's not a bad error message -- just says that GHC can't make progress on e.g. `C a` because it appears to be looping).

I do think that we could do a better job educating the Haskelling public that `UndecidableInstances` is safe to use.

**Comment by noughtmare** [comment]
I think there is a second problem unrelated to compiler termination, namely that instances of the form `instance C a` preclude other non-overlapping instances. For a class like `Ord` as used in the linked stackoverflow answer, a newtype wrapper is a good solution. However, in the abstract example in this issue it is much less appealing to reach for newtypes.

**Comment by silky** [comment]
Interesting @goldfirere ; I certainly didn't know that; and yes, the phrase `UndecidableInstances` concerned me enough to not even bother to investigate turning it on 😅

**Comment by goldfirere** [comment]
Maybe we should make it so that every error message that suggests `UndecidableInstances` includes

```
NB: UndecidableInstances means that GHC's constraint solver
might not terminate, but it does not compromise the safety
or correctness of compiled programs.
```

That's a bit wordy, but maybe its educational value is high enough to be worth the price. If you can get the sentiment expressed more concisely, please help!

**Comment by silky** [comment]
Is there a good short-hand or intuition for understanding how I would use undecidable instances to _force_ it to non-terminate? i guess the other risk of undecidable-instances is that it's global for the file, and not specific to the instance that i'm creating just for that moment (i think i've seen this discussed somewhere).

If it could easily be described in a few words, then i suppose something like "Non-termination occurs when ...; allowing Undecidable instances bypasses this safety check, at the risk of compilation running indefinitely.", or some better-worded sentence

**Comment by goldfirere** [comment]
There are specific checks that we could disable for specific instances, yes. But the real effect of `UndecidableInstances` is that it can cause non-termination in any downstream module. That is, if any module you import (transitively) uses `UndecidableInstances`, your module might not terminate during type-checking.

It's also hard to characterize when non-termination occurs -- just like it would be in term-level programming. I can offer an example of non-termination, though:

```hs
class C a where
  method :: a -> ()

instance C (Maybe a) => C a where
  method _ = ()

foo x = method x
```

This errors with a message saying `Reduction stack overflow` (and more information), which I think is somewhat reasonable. (Maybe the word "reduction" doesn't help.)

This is a fairly obvious case. But we can easily imagine a network of instance declarations all of which work together to form a loop. And there are other cases involving functional dependencies. It's a hard thing to pin down, as non-termination often is.

Given this -- and maybe inspired by the example -- do you see a good way forward?

**Comment by silky** [comment]
I see.

Thanks for clarifying @goldfirere .

I'm not sure I see a good way forward. I guess you're saying it has something to do with creating a recursive loop; so maybe the error could hint at the essence of that somehow; "It looks like we could get into a loop resolving instances; you can break the loop via ... or if you're sure you're not introducing a loop try ..."

But it does feel a bit verbose?
